File: src/Cistern.Linq.FSharp/TypedEnumerables.fs
module private Cistern.Linq.FSharp.TypedEnumerables

open Cistern.Linq.ChainLinq.Optimizations

[<Struct; NoComparison; NoEquality>]
type FSharpListEnumerator<'T> =
    val mutable state : list<'T>
    val mutable current : 'T

    new (l:list<'T>) = {
        state = l
        current = Unchecked.defaultof<'T>
    }

    interface System.Collections.Generic.IEnumerator<'T> with
        member this.Current: 'T = this.current

    interface System.IDisposable with
        member __.Dispose() = ()

    interface System.Collections.IEnumerator with
        member this.MoveNext () = 
            match this.state with
            | [] -> false
            | hd :: tl ->
                this.current <- hd
                this.state <- tl
                true

        member __.Current  = raise (System.NotImplementedException())
        member __.Reset () = raise (System.NotImplementedException())

[<Struct; NoComparison; NoEquality>]
type FSharpListEnumerable<'T>(lst:list<'T>) =
    interface ITypedEnumerable<'T, FSharpListEnumerable<'T>, FSharpListEnumerator<'T>> with
        member __.Source = upcast lst
        member __.TryGetSourceAsSpan _ = false
        member __.TryLength = System.Nullable lst.Length // is this O(1)??
        member __.GetEnumerator () = new FSharpListEnumerator<'T>(lst)
        member __.TrySkip (count, enumerable:byref<_>) =
            let rec skip count = function
            | [] -> FSharpListEnumerable []
            | lst when count = 0 -> FSharpListEnumerable lst
            | _ :: tl -> skip (count-1) tl
            enumerable <- skip count lst
            true

File: src/Cistern.Linq.System/TypedEnumerables.cs
using Cistern.Linq.ChainLinq.Optimizations;
using System;
using System.Collections.Generic;

namespace Cistern.Linq.Immutable
{
    struct LinkedListEnumerable<T>
        : ITypedEnumerable<T, LinkedList<T>.Enumerator>
    {
        private LinkedList<T> source;
        public LinkedListEnumerable(LinkedList<T> source) => this.source = source;

        public IEnumerable<T> Source => source;

        public int? TryLength => source.Count;

        public LinkedList<T>.Enumerator GetEnumerator() => source.GetEnumerator();

        public bool TryGetSourceAsSpan(out ReadOnlySpan<T> readOnlySpan)
        {
            readOnlySpan = default;
            return false;
        }

        public bool TryLast(out T result)
        {
            var last = source.Last;
            if (last == null)
            {
                result = default;
                return false;
            }
            result = last.Value;
            return true;
        }

    }

    struct HashSetEnumerable<T>
        : ITypedEnumerable<T, HashSet<T>.Enumerator>
    {
        private HashSet<T> source;
        public HashSetEnumerable(HashSet<T> source) => this.source = source;

        public IEnumerable<T> Source => source;
        public int? TryLength => source.Count;

        public HashSet<T>.Enumerator GetEnumerator() => source.GetEnumerator();
        public bool TryGetSourceAsSpan(out ReadOnlySpan<T> readOnlySpan)
        {
            readOnlySpan = default;
            return false;
        }

        public bool TryLast(out T result)
        {
            result = default;
            return false;
        }
    }

    struct StackEnumerable<T>
        : ITypedEnumerable<T, Stack<T>.Enumerator>
    {
        private Stack<T> source;
        public StackEnumerable(Stack<T> source) => this.source = source;

        public IEnumerable<T> Source => source;
        public int? TryLength => source.Count;

        public Stack<T>.Enumerator GetEnumerator() => source.GetEnumerator();
        public bool TryGetSourceAsSpan(out ReadOnlySpan<T> readOnlySpan)
        {
            readOnlySpan = default;
            return false;
        }

        public bool TryLast(out T result)
        {
            result = default;
            return false;
        }
    }

    struct QueueEnumerable<T>
        : ITypedEnumerable<T, Queue<T>.Enumerator>
    {
        private Queue<T> source;
        public QueueEnumerable(Queue<T> source) => this.source = source;

        public IEnumerable<T> Source => source;
        public int? TryLength => source.Count;

        public Queue<T>.Enumerator GetEnumerator() => source.GetEnumerator();
        public bool TryGetSourceAsSpan(out ReadOnlySpan<T> readOnlySpan)
        {
            readOnlySpan = default;
            return false;
        }

        public bool TryLast(out T result)
        {
            result = default;
            return false;
        }
    }

    struct SortedSetEnumerable<T>
        : ITypedEnumerable<T, SortedSet<T>.Enumerator>
    {
        private SortedSet<T> source;
        public SortedSetEnumerable(SortedSet<T> source) => this.source = source;

        public IEnumerable<T> Source => source;
        public int? TryLength => source.Count;

        public SortedSet<T>.Enumerator GetEnumerator() => source.GetEnumerator();
        public bool TryGetSourceAsSpan(out ReadOnlySpan<T> readOnlySpan)
        {
            readOnlySpan = default;
            return false;
        }

        public bool TryLast(out T result)
        {
            result = default;
            return false;
        }
    }
}
